### PR TITLE
Generate pull request preview links

### PR DESCRIPTION
**Pull request by aswanson-nr**
# Summary
* Adds a workflow to generate links for pull request previews
* Adds script to create the links using the github API
* Moved some helper functions out of quickstart validation and into the general `helpers.ts`

## Testing
- [1 quickstart](https://github.com/aswanson-nr/newrelic-quickstarts/pull/23)
- [3 quickstarts](https://github.com/aswanson-nr/newrelic-quickstarts/pull/25)
- [no quickstarts](https://github.com/aswanson-nr/newrelic-quickstarts/pull/24)

### CODE REVIEW

File: utils/create-preview-links.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ * Executes this script to create a pull request comment with preview links for each quickstart in the pull request
+ * Requires the `GITHUB_TOKEN` environment variable to be set to a Github access token
+ * @param prURL - The Github API URL to the pull request
+ * @param prNumber - The Github pull request number
+ * @param token - A Github access token
+ */
+export const generatePreviewComment = async (
+  prURL: string | undefined,
+  prNumber: string | undefined,
+  token: string | undefined
```
Comment (zstix) [inline utils/create-preview-links.ts:71]: Totally optional, but you could use Typescript's [optional parameter syntax](https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters) here. This is equivalent:

```suggestion
  prURL?: string,
  prNumber?: string,
  token?: string
```

File: utils/create-preview-links.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  return `${heading}${markdownLinks}`;
+};
+
+/**
+ * Executes this script to create a pull request comment with preview links for each quickstart in the pull request
+ * Requires the `GITHUB_TOKEN` environment variable to be set to a Github access token
+ * @param prURL - The Github API URL to the pull request
+ * @param prNumber - The Github pull request number
+ * @param token - A Github access token
```
Comment (zstix) [inline utils/create-preview-links.ts:66]: If we're using [JSDoc syntax](https://jsdoc.app/tags-param.html), should we denote that these parameters are optional? Or should we just let Typescript handle that stuff?

Here's what that would look like:

```suggestion
 * @param [prURL] - The Github API URL to the pull request
 * @param [prNumber] - The Github pull request number
 * @param [token] - A Github access token
```

I don't feel strongly either way. Feel free to keep it as-is.

File: utils/helpers.ts [hunk 3]
```diff
@@ -184,3 +184,74 @@ export const cleanQuickstartName = (str: string): string =>
...
+/**
+ * Gets the unique base quickstart directory from a given file path.
+ * e.g. filePath: 'quickstarts/python/aiohttp/alerts/ApdexScore.yml' + targetChild: 'alerts' = 'python/aiohttp'.
+ * @param {String} filePath - Full file path of a file in a quickstart.
+ * @param {String} targetChild - Node in file path that should be preceded by a base quickstart directory.
+ * @return {String} Node in file path of the quickstart.
+ */
+const getQuickstartNode = (
+  filePath: string,
+  targetChild: string | undefined
```
Comment (zstix) [inline utils/helpers.ts:197]: Potential for optional parameter syntax (if you want):

```suggestion
  targetChild?: string
```

File: utils/helpers.ts [hunk 4]
```diff
@@ -184,3 +184,74 @@ export const cleanQuickstartName = (str: string): string =>
     .replace(/\s+/g, '-')
     .replace(/-+/, '-')
     .replace(/[^a-z0-9-]/g, '');
+
+/**
+ * Gets the unique base quickstart directory from a given file path.
+ * e.g. filePath: 'quickstarts/python/aiohttp/alerts/ApdexScore.yml' + targetChild: 'alerts' = 'python/aiohttp'.
+ * @param {String} filePath - Full file path of a file in a quickstart.
+ * @param {String} targetChild - Node in file path that should be preceded by a base quickstart directory.
+ * @return {String} Node in file path of the quickstart.
```
Comment (zstix) [inline utils/helpers.ts:193]: It looks like we're omitting the type information in the JSDoc comments in the other files in this PR. We might want to keep this file consistent:

```suggestion
 * @param filePath - Full file path of a file in a quickstart.
 * @param [targetChild] - Node in file path that should be preceded by a base quickstart directory.
 * @return Node in file path of the quickstart.
```

File: utils/helpers.ts [hunk 5]
```diff
@@ -184,3 +184,74 @@ export const cleanQuickstartName = (str: string): string =>
...
+ */
+const getQuickstartNode = (
+  filePath: string,
+  targetChild: string | undefined
+): string => {
+  const splitFilePath = filePath.split('/');
+
+  const baseQuickstartDirectoryIndex =
+    splitFilePath.findIndex((path) => path === targetChild) - 1;
```
Comment (zstix) [inline utils/helpers.ts:202]: Could we use [`Array.prototype.indexOf()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf) here?
```suggestion
  const baseQuickstartDirectoryIndex = splitFilePath.indexOf(targetChild);
```

File: utils/helpers.ts [hunk 6]
```diff
@@ -184,3 +184,74 @@ export const cleanQuickstartName = (str: string): string =>
...
+    );
+    indexCounter--;
+  }
+  return uniqueQuickstartDirectory;
+};
+
+/**
+ * Identifies where in a given file path to look for a quickstart directory.
+ * @param {String} filePath - Full file path of a file in a quickstart.
+ * @return {String|undefined} Called function with arguments to determine the quickstart of a given file path.
```
Comment (zstix) [inline utils/helpers.ts:219]: Same comment as above. Totally optional.
```suggestion
 * @param filePath - Full file path of a file in a quickstart.
 * @return Called function with arguments to determine the quickstart of a given file path.
```

**Review by zstix — APPROVED** [review]
This looks great!

I have a few optional comments, but nothing I would block the PR for.

File: .github/workflows/preview-links.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        id: links
+        run: |
+          yarn create-preview-links ${{ github.event.pull_request.url }} ${{ github.event.pull_request.number }}
+
+      - name: Comment 
+        run: |
+          if [[ '${{ steps.links.outputs.comment }}' != '' ]]; then
+              echo "Detected comment, posting to pull request"
+          gh pr comment ${{github.event.pull_request.number}} --body ${{ steps.links.outputs.comment }}
+          fi
```
[the remark this answers is not in the document]
Comment (aswanson-nr) [reply .github/workflows/preview-links.yml]: This is much more simple! love it

File: utils/create-preview-links.ts (hunk 1 above)
Comment (aswanson-nr) [reply utils/create-preview-links.ts]: Done!

File: utils/create-preview-links.ts (hunk 2 above)
Comment (aswanson-nr) [reply utils/create-preview-links.ts]: Done!

File: utils/helpers.ts (hunk 3 above)
Comment (aswanson-nr) [reply utils/helpers.ts]: Done!

File: utils/helpers.ts (hunk 4 above)
Comment (aswanson-nr) [reply utils/helpers.ts]: Done!

File: utils/helpers.ts (hunk 5 above)
Comment (aswanson-nr) [reply utils/helpers.ts]: Looks like it to me! I updated the function

File: utils/helpers.ts (hunk 6 above)
Comment (aswanson-nr) [reply utils/helpers.ts]: Done!

**Review by zstix — APPROVED** [review]

**Comment by jpvajda** [comment]
😍 😍 😍

**Pull request merged**